keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -340,17 +341,80 @@ Con:
  - we must add a note to the documentation that if such Service is created then it may break things after a rollback - it depends on the cloud provider implementation
  - the update of such Services shall be possible even if the feature flag is OFF. This is to prepare for the next release when the feature flag is removed from the create path, too, and after a rollback to the first release the update of existing Service objects must be possible
- - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. It does not mean that the CPI implementations have to accept such Services. It means that the CPI implementations shall be able to process such Services. The CPI implementation can still validate the Service objects to check whether any of those has mixed protocol configuration and if the CPI cannot support such Service for any reasons the CPI shall indicate that clearly to the user, and the CPI should not create Cloud LB resources for the "accepted part" of the Service. The point is, that the system (Service - CPI implementation - cloud LB) must be in a consitent state in the end, and the user must have meaningful information about that state. For example, if the Service is not accepted by the CPI implementation this fact shall be available for the user, either as a new field in `Service.status.loadBalancer` or as a new Event (or both).
+ - the CPI implementations shall be prepared to deal with Services with mixed protocol configurations. Either via supporting such Service definitions, or clearly indicating to the users that the Service could not be processed as specified. As we can see from our analysis some CPIs support other protocols than TCP and UDP in the Service definitions, while others support only TCP and UDP. That is the term "mixed protocol support" does not always mean that all possible protocol values are supported by a CPI. For this reason a nicely behaving CPI shall 
+ - indicate clearly to the user what ports with what protocols have been opened on the LB
+ - preferably not create any Cloud LB resources if the Service definition contains unsupported protocols.
+ 
+ In order to provide a way for the CPI to indicate port statuses to the user we would add the following new `portStatus` list of port statuses to `Service.status.loadBalancer`, so the CPI can indicate the status of the LB:
+
+```json
+"io.k8s.api.core.v1.PortStatus": {
+  "description": "PortStatus contains details for the current status of this port.",
+  "properties": {
+    "port": {
+      "description": "Port number",
+      "format": "int32",
+      "type": "integer"
+    },
+    "protocol": {
+      "description": "The protocol for this port",
+      "type": "string"
+    },
+    "name": {
+      "description": "The name of this port within the service.",
+      "type": "string"
+    },
+    "ready": {
```

Comment:
Are there more states that a single port or set of ports might be in? I'm wondering if bool is right or enum or even Conditions.  Bool is simple, maybe good enough...

Reply:
I have been thinking about it, but could not find more states for the time being.